### PR TITLE
add histogram LGDO

### LINKED ISSUE #15: Implement histogram LH5 specification
Serialization/deserialization of [hist.Hist](hist.readthedocs.io) objects to LH5.

Specification: https://legend-exp.github.io/legend-data-format-specs/dev/hdf5/#Histograms

### PR DESCRIPTION
**Pull request by ManuelHu**
Some open questions:
* this currently has a hard import on the `hist` package, which could also be a strictly optional dependency. see also #48 
* I added an error message when trying to write a histogram with any `wo_mode` that might append. I am unsure if this is the nice way to do it...
* For `Histogram`, not being able to add/remove fields is a hard requirement - otherwise it cannot be read back as a histogram. But overriding all the custom functionality of `Struct` is not nice.  Maybe, having a base class for `Struct`, but without a clear way for the user to add/remove fields would make sense?
* Should a `HistogramAxis` accept attrs?
* Should `lgdo.types.HistogramAxis` be moved to `lgdo.types.Histogram.Axis` (i.e. as nested type, as the axis type does not make sense as a standalone LGDO?)

fixes #15

### CODE REVIEW

File: src/lgdo/types/histogram.py
Comment (gipert) [inline src/lgdo/types/histogram.py:114]:
```suggestion
                msg = "flow bins of hist.Hist cannot be represented"
```

**Comment by gipert** [comment]
Looks great Manuel! Some comments:

- we should also support variable-binning histograms
- I would ignore the "problem" with mutable `Struct`s for now 
- I think going for `lgdo.types.Histogram.Axis` is a good idea
- could you document a bit better if data is copied or actually viewed?

@oschulz:

- https://legend-exp.github.io/legend-data-format-specs/dev/hdf5/#Histograms does not mention histograms with variable binning, but should be straightforward to have `binedges` as an `array<1>{real}`.
- could you provide an LH5 file with a Histogram object serialized with Julia?

**Review by gipert — CHANGES_REQUESTED** [review]

**Comment by ManuelHu** [comment]
* conversion `hist.Hist` -> `np.ndarray` is (mostly) zero-copy
* conversion `np.ndarray/Histogram` -> `hist.Hist` (i.e. in `view_as("hist")` is _not_ just a view, but performs a full copy! Unfortunately there is no way to change it: https://github.com/scikit-hep/boost-histogram/issues/421

**Comment by oschulz** [comment]
> https://legend-exp.github.io/legend-data-format-specs/dev/hdf5/#Histograms does not mention histograms with variable binning, but should be straightforward to have binedges as an array<1>{real}.

Yes, we should absolutely allow that.

> could you provide an LH5 file with a Histogram object serialized with Julia?

I'll add a test file to legend-testdata, give me a few days ...

**Comment by ManuelHu** [comment]
Variable binning works now (with `array<1>{real}` as `binedges`). I can also prepare a PR to the data specification in the next days.
There is also some logic that converts an edge array to a range object, if this is possible (i.e. it has a uniform binning). I Am not sure if we want to keep that behavior...

but please do not merge yet, I still have one change in mind. I still want to change one possible parameter type of the binning, as now it is not so nice.

**Comment by oschulz** [comment]
> I can also prepare a PR to the data specification in the next days.

Can you also add a test file to the legend-testdata repo via a PR there, so we can test on the Julia side before we merge here and on legend-testdata?

**Comment by ManuelHu** [comment]
@oschulz see https://github.com/legend-exp/legend-testdata/pull/18
I now uploaded a test file with two 2D histograms, one with range axes, and one with variable binned axes.

File: src/lgdo/types/histogram.py
Comment (gipert) [inline src/lgdo/types/histogram.py:157]:
```suggestion
        weights: hist.Hist | NDArray,
```

we should use more generic types in function signatures (see [remarks here](https://learn.scientific-python.org/development/guides/mypy/#tips-for-good-types)). Could you quickly go through all your function signatures and fix this?

**Review by gipert — CHANGES_REQUESTED** [review]

**Comment by gipert** [comment]
This PR seems good to go from my point of view, anything else @ManuelHu?

**Comment by ManuelHu** [comment]
I also think it is ready, no particularly missing things come to my mind.

**Pull request merged**
